src/main/java/com/uw/controller/TraineeController.java

```diff
@@ -0,0 +1,361 @@
+            newUsername = traineeToUpdate.getUser().getUsername();
+        }
+        traineeToUpdate.getUser().setUsername(newUsername);
+
+        traineeService.updateTrainee(traineeToUpdate);
+        return ResponseEntity.ok().body(traineeToUpdate);
+    }
+
+    @DeleteMapping("/{username}")
+    public ResponseEntity<?> delete(@PathVariable("username") String username,
+                                    @RequestHeader(value = "Authorization") String auth){
+        String[] credentials = AuthService.decodeCredentials(auth);
+        if (credentials.length != 2) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid Authorization header");
+        }
+
+        String providedUsername = credentials[0];
+        String providedPassword = credentials[1];
+
+        // Autenticar el usuario
+        boolean isAuthenticated = authService.authentication(providedUsername, providedPassword);
+        if (!isAuthenticated) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid credentials");
+        }
+
+        if (!authService.isTrainee(providedUsername, providedPassword)) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Trainee not found");
+        }
+
+        Trainee traineeToDelete = traineeService.findTraineeByUsername(username);
+        if (traineeToDelete == null) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Trainee not found");
+        }
+
+        traineeService.deleteTraineeByUserName(username);
+
+        return ResponseEntity.status(HttpStatus.OK).body("Successfully deleted");
+    }
+
+    @GetMapping("/{username}/not-assigned-trainers")
+    public ResponseEntity <?> getNotAssignedActiveTrainers(
+            @PathVariable("username") String username,
+            @RequestHeader(value = "Authorization") String auth) {
+
+        String[] credentials = AuthService.decodeCredentials(auth);
+
+        if (credentials.length != 2) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid Authorization header");
+        }
+
+        String providedUsername = credentials[0];
+        String providedPassword = credentials[1];
+
+        // Autenticar el usuario
+        boolean isAuthenticated = authService.authentication(providedUsername, providedPassword);
+        if (!isAuthenticated) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid credentials");
+        }
+
+        if (!authService.isTrainee(providedUsername, providedPassword)) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Trainee not found");
+        }
+
+        Trainee trainee = traineeService.findTraineeByUsername(username);
+
+        List<Trainer> trainers = trainerService.findAll();
+        List<Trainer> results = new ArrayList<>();
+        for (Trainer trainer : trainers){
```

Comment:
To avoid O(n-square) compexity probably you will need to use Map structure instead of List , leave as is but take into account this tips that could impact performance